libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+        candidates[0] = oaptr;
+        num_candidates = 1;
+        break;
+      } else if (stage->words[w_idx] && stage->words[w_idx][0] == '-' && (oaptr->flags & (CLI_CMD_HYPHENATED_OPTION))) {
+        candidates[0] = oaptr;
+        num_candidates = 1;
+        break;
+      } else if (!stage->words[w_idx] || (oaptr->flags & CLI_CMD_ARGUMENT) ||
+                 !strncasecmp(oaptr->name, stage->words[w_idx], strlen(stage->words[w_idx]))) {
+        candidates[num_candidates++] = oaptr;
+      }
+      if (oaptr->flags & CLI_CMD_ARGUMENT) {
+        break;
+      }
+    }
+
+    /*
+     * Iterate over the list of candidates for this word.  There are several early exit cases to consider:
+     * - If we have no candidates then we're done - any remaining words must be processed by the command callback
+     * - If we have more than one candidate evaluating for execution punt hard after complaining.
+     * - If we have more than one candidate and we're not at end-of-line (
+     */
+    if (num_candidates == 0) break;
+    if ((num_candidates > 1) && ((lastchar == '\0') || (w_idx < (stage->num_words - 1)))) {
+      stage->error_word = stage->words[w_idx];
+      stage->status = CLI_AMBIGUOUS;
+      cli_error(cli, "Ambiguous option/argument for command %s", stage->command->command);
+      return;
+    }
+
+    /*
+     * So now we could have one or more candidates.  We need to call get help/completions *only* if this is the
+     * 'last-word'
+     * Remember that last word for optinal arguments is last or next to last....
+     */
+    if (lastchar != '\0') {
+      int called_comphelp = 0;
+      for (c_idx = 0; c_idx < num_candidates; c_idx++) {
+        oaptr = candidates[c_idx];
+
+        // need to know *which* word we're trying to complete for optional_args, hence the difference calls
+        if (((oaptr->flags & (CLI_CMD_OPTIONAL_FLAG | CLI_CMD_ARGUMENT)) && (w_idx == (stage->num_words - 1))) ||
+            ((oaptr->flags & (CLI_CMD_OPTIONAL_ARGUMENT | CLI_CMD_HYPHENATED_OPTION)) &&
+             (w_idx == (stage->num_words - 1)))) {
+          cli_get_optarg_comphelp(cli, oaptr, comphelp, num_candidates, lastchar, stage->words[w_idx],
+                                  stage->words[w_idx]);
+          called_comphelp = 1;
+        } else if ((oaptr->flags & CLI_CMD_OPTIONAL_ARGUMENT) && (w_idx == (stage->num_words - 2))) {
+          cli_get_optarg_comphelp(cli, oaptr, comphelp, num_candidates, lastchar, stage->words[w_idx],
+                                  stage->words[w_idx + 1]);
+          called_comphelp = 1;
+        }
+      }
+      // if we were 'end-of-word' and looked for completions/help, return to user
+      if (called_comphelp) {
+        stage->status = CLI_OK;
+        return;
+      }
+    }
+
+    // set some values for use later - makes code much easier to read
+    if (((oaptr->flags & (CLI_CMD_OPTIONAL_FLAG | CLI_CMD_ARGUMENT)) && (w_idx == (stage->num_words - 1))) ||
```

Comment:
I think it's possible for `oaptr` to be NULL at this point

Reply:
Good catch - this if clause needs to be after the next three lines - this block is only hit *if* we have exactly one candidate, to which oaptr points.